code/modules/antagonists/mind_flayer/flayer_power.dm

```diff
@@ -0,0 +1,263 @@
+
+/datum/spell/flayer/self/augment_menu/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "AugmentMenu", name)
+		ui.set_autoupdate(FALSE)
+		ui.open()
+
+/datum/spell/flayer/self/augment_menu/ui_act(action, list/params, datum/tgui/ui)
+	var/mob/user = ui.user
+	if(user.stat)
+		return
+
+	switch(action)
+		if("purchase")
+			var/path = text2path(params["ability_path"])
+			on_purchase(user, path)
+			update_static_data(ui.user)
+
+// Takes in a category name and grabs the paths of all the spells/passives specific to that category. Used for TGUI
+/datum/antagonist/mindflayer/proc/get_powers_of_category(category)
+	var/list/list/powers = list()
+	for(var/path in ability_list)
+		if(is_path_spell(path))
+			var/datum/spell/flayer/spell = path
+			if(spell.category == category)
+				powers += list(list(
+					"name" = spell.name,
+					"desc" = spell.desc,
+					"max_level" = spell.max_level,
+					"cost" = spell.base_cost,
+					"stage" = spell.stage,
+					"ability_path" = spell.type
+				))
+		else
+			var/datum/mindflayer_passive/passive = path
+			if(passive.category == category)
+				powers += list(list(
+					"name" = passive.name,
+					"desc" = passive.purchase_text,
+					"max_level" = passive.max_level,
+					"cost" = passive.base_cost,
+					"stage" = passive.stage,
+					"ability_path" = passive.type
+				))
+
+	return powers
+
+/datum/antagonist/mindflayer/proc/build_ability_tabs()
+	var/list/list/ability_tabs = list()
+	for(var/category in category_stage)
+		ability_tabs += list(list(
+			"category_name" = category,
+			"category_stage" = category_stage[category],
+			"abilities" = get_powers_of_category(category)
+			))
+	return ability_tabs
+
+/datum/spell/flayer/self/augment_menu/ui_data(mob/user)
+	var/list/list/data = list()
+	var/list/known_abilities = list()
+	data["usable_swarms"] = flayer.usable_swarms
+	for(var/datum/mindflayer_passive/passive in flayer.powers)
+		known_abilities += list(list(
+			"name" = passive.name,
+			"current_level" = passive.level,
+			"max_level" = passive.max_level,
+			"cost" = passive.current_cost,
+			"upgrade_text" = passive.upgrade_info,
+			"ability_path" = passive.type
+		))
+	for(var/datum/spell/flayer/spell in flayer.powers)
+		known_abilities += list(list(
+			"name" = spell.name,
+			"current_level" = spell.level,
+			"max_level" = spell.max_level,
+			"cost" = spell.current_cost,
+			"upgrade_text" = spell.upgrade_info,
+			"ability_path" = spell.type
+		))
+	data["known_abilities"] = known_abilities
+	return data
+
+/datum/spell/flayer/self/augment_menu/ui_static_data(mob/user)
+	var/list/list/static_data = list()
+	static_data["ability_tabs"] = flayer.build_ability_tabs()
+	return static_data
+
+// Given a path, return TRUE if the path is a mindflayer spell, or FALSE otherwise. Only used to sort passives from spells.
+/datum/antagonist/mindflayer/proc/is_path_spell(path)
+	return ispath(path, /datum/spell)
```

Comment:
Could this be a define/global proc? It definitely doesn't need to be a flayer proc